### ISSUE TITLE
The refactored code on main is giving weird accuracy when training Qwen3VL-8B, and the entropy keeps going up abnormally during training.

### ISSUE THREAD

**Issue by DearFishi** [verl-project/verl#6382]
### System Info

----------Python Info----------
Version      : 3.11.14
Compiler     : GCC 14.3.0
Build        : ('main', 'Oct 22 2025 22:46:25')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 25.2
Directory    : /opt/conda/lib/python3.11/site-packages/pip
vllm         : 0.13.0
sglang       : not found.
ray          : 2.53.0
torch        : 2.9.0
----------verl Info-----------
Version      : 0.8.0.dev
Directory    : /workspace/verl/verl
Commit Hash  : ed9b914fac3d34379259a7beb32e8a9145027de4
----------Platform Info----------
Platform     : Linux-4.14.0_1-0-0-52-x86_64-with-glibc2.35
system       : Linux
release      : 4.14.0_1-0-0-52
version      : #5 SMP Thu Sep 14 16:47:18 CST 2023
----------Environment----------
CUDA Runtime : 12.8
CUDA Compiler : Cuda compilation tools, release 12.8, V12.8.93
----------System Info----------
CPU Memory      : 1007.06 GB
GPU Count       : 8
GPU 1   Type    : NVIDIA A800-SXM4-80GB
GPU 1   Memory  : 80.00 GB
GPU 2   Type    : NVIDIA A800-SXM4-80GB
GPU 2   Memory  : 80.00 GB
GPU 3   Type    : NVIDIA A800-SXM4-80GB
GPU 3   Memory  : 80.00 GB
GPU 4   Type    : NVIDIA A800-SXM4-80GB
GPU 4   Memory  : 80.00 GB
GPU 5   Type    : NVIDIA A800-SXM4-80GB
GPU 5   Memory  : 80.00 GB
GPU 6   Type    : NVIDIA A800-SXM4-80GB
GPU 6   Memory  : 80.00 GB
GPU 7   Type    : NVIDIA A800-SXM4-80GB
GPU 7   Memory  : 80.00 GB
GPU 8   Type    : NVIDIA A800-SXM4-80GB
GPU 8   Memory  : 80.00 GB

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. docker: pytorch/pytorch:2.9.1-cuda12.8-cudnn9-devel
2. cd verl && pip install -e .
3. pip install vllm=0.13.0
4. pip install qwen-vl-utils && pip install mathruler
5. script: 
```set -x
ENGINE=${1:-vllm}

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=/workspace/geo3k/train.parquet \
    data.val_files=/workspace/geo3k/test.parquet \
    data.train_batch_size=512 \
    data.max_prompt_length=1024 \
    data.max_response_length=2048 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.image_key=images \
    actor_rollout_ref.model.path=/workspace/Qwen3-VL-8B-Instruct \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.optim.lr_warmup_steps=10 \
    actor_rollout_ref.actor.optim.weight_decay=0.1 \
    actor_rollout_ref.actor.optim.clip_grad=1.0 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.model.use_fused_kernels=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.model.enable_gradient_checkpointing=False \
    actor_rollout_ref.actor.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=10 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=4 \
    actor_rollout_ref.rollout.name=$ENGINE \
    actor_rollout_ref.rollout.max_model_len=12800 \
    +actor_rollout_ref.rollout.enable_sleep_mode=True \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.7 \
    actor_rollout_ref.rollout.dtype="float16" \
    ++actor_rollout_ref.rollout.top_p=0.9 \
    actor_rollout_ref.rollout.enable_prefix_caching=True \
    actor_rollout_ref.rollout.enable_chunked_prefill=True \
    actor_rollout_ref.rollout.enforce_eager=True \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.rollout.n=5 \
    actor_rollout_ref.rollout.val_kwargs.top_p=0.7 \
    actor_rollout_ref.rollout.val_kwargs.do_sample=True \
    actor_rollout_ref.rollout.val_kwargs.n=1 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=20 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.rollout_correction=null \
    trainer.critic_warmup=0 \
    trainer.logger='["console", "wandb"]' \
    trainer.val_before_train=False \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=5 \
    trainer.total_epochs=15 $@
```

### Expected behavior

<p align="center">
  <img src="https://github.com/user-attachments/assets/c612eb0e-4543-4dd4-950e-7ccd085ca383" width="32%" />
  <img src="https://github.com/user-attachments/assets/75eae6af-dfc7-4623-92b0-73b79eec0b83" width="32%" />
  <img src="https://github.com/user-attachments/assets/84fc2e56-77f6-4ad2-896f-9b7298d2bafc" width="32%" />
</p>

<p align="center">
  <img src="https://github.com/user-attachments/assets/85a99e9e-1463-4d6e-b187-d0403d398806" width="48%" />
  <img src="https://github.com/user-attachments/assets/95aa4b20-cbc1-4f0c-9d7e-3bfd3f92fe3d" width="48%" />
</p>

**Comment by Kirrito-k423** [comment]
Thank you for the detailed bug report with profiling data.

**Initial analysis of the entropy behavior:**

The abnormally rising entropy during training typically indicates one of the following issues:

1. **Learning rate too high** —  for Qwen3-VL-8B could be aggressive. The entropy rise suggests the policy is becoming more uniform (less confident), often a sign of over-optimization.

2. **KL loss coefficient** —  is relatively high. This pushes the policy away from the reference model aggressively, which can cause entropy to spike if the advantage estimator is noisy.

3. ** Advantage estimator instability** — With  and GRPO, the advantage estimation can have high variance early in training, causing erratic policy updates.

**Recommended diagnostic steps:**

1. Check if the issue occurs with a **lower learning rate** (e.g.,  or )
2. Reduce  to  to see if entropy behavior stabilizes
3. Monitor  in the logs — if it's very high (>5), the policy is moving too far from reference

Could you share the training log from the first few steps? Specifically looking for:
- 
- 
- 

The images you shared show accuracy dropping to near-zero while entropy climbs — this pattern suggests the policy is collapsing to a mode that produces consistently wrong answers.

**Comment by DearFishi** [comment]
> Thank you for the detailed bug report with profiling data.
> 
> **Initial analysis of the entropy behavior:**
> 
> The abnormally rising entropy during training typically indicates one of the following issues:
> 
> 1. **Learning rate too high** —  for Qwen3-VL-8B could be aggressive. The entropy rise suggests the policy is becoming more uniform (less confident), often a sign of over-optimization.
> 
> 2. **KL loss coefficient** —  is relatively high. This pushes the policy away from the reference model aggressively, which can cause entropy to spike if the advantage estimator is noisy.
> 
> 3. ** Advantage estimator instability** — With  and GRPO, the advantage estimation can have high variance early in training, causing erratic policy updates.
> 
> **Recommended diagnostic steps:**
> 
> 1. Check if the issue occurs with a **lower learning rate** (e.g.,  or )
> 2. Reduce  to  to see if entropy behavior stabilizes
> 3. Monitor  in the logs — if it's very high (>5), the policy is moving too far from reference
> 
> Could you share the training log from the first few steps? Specifically looking for:
> - 
> - 
> - 
> 
> The images you shared show accuracy dropping to near-zero while entropy climbs — this pattern suggests the policy is collapsing to a mode that produces consistently wrong answers. 


- commit: https://github.com/verl-project/verl/commit/ed9b914fac3d34379259a7beb32e8a9145027de4 train log: 
[train_ed9b914.log](https://github.com/user-attachments/files/27982245/train_ed9b914.log)

- I previously trained with commit [b2c489c](https://github.com/verl-project/verl/commit/b2c489c3b88a78f6e2abe943e22ac2518cb175f6), and the training results looked relatively normal. The training script, data, and model were all unchanged, but the hardware used was not A800.
<p align="center">
    <img src="https://github.com/user-attachments/assets/ccc56ded-0655-47b3-af95-35aa87a41b7d"  width="32%" />
    <img src="https://github.com/user-attachments/assets/98d0f95a-e1b4-4fe6-bf64-33b8274742e3" width="32%" />
    <img src="https://github.com/user-attachments/assets/8a3330a5-aaa6-4ac0-a0f1-b006bba8dd8e" width="32%"/>
</p>
<p align="center">
    <img src="https://github.com/user-attachments/assets/7de83e9d-d52b-4d99-88a2-7aaef6573556" width="48%" />
    <img src="https://github.com/user-attachments/assets/df651894-ef40-4770-ae20-0ede9e568ce0" width="48%" />
</p>
